src/store/SwapMetaIn.cc

```diff
@@ -0,0 +1,361 @@
+               Debug::Extra << "expected: " << emem.urlXXX() <<
+               Debug::Extra << "found:    " << storedUrl);
+        throw TextException("URL mismatch", Here());
+    }
+}
+
+/// deserializes STORE_META_VARY_HEADERS swap meta field
+static SBuf
+UnpackNewSwapMetaVaryHeaders(const SwapMetaView &meta, const StoreEntry &entry)
+{
+    Assure(meta.type == STORE_META_VARY_HEADERS);
+    SBuf rawVary(static_cast<const char *>(meta.rawValue), meta.rawLength);
+    // entries created before SBuf-based Vary may include string terminator
+    static const SBuf nul("\0", 1);
+    rawVary.trim(nul, false, true);
+
+    const auto &knownVary = entry.mem().vary_headers;
+    if (knownVary.isEmpty())
+        return rawVary; // new Vary (that we cannot validate)
+
+    if (knownVary == rawVary)
+        return SBuf(); // OK: no new Vary
+
+    throw TextException("Vary mismatch", Here());
+}
+
+/// deserializes entry metadata size from the given buffer
+/// \retval total swap metadata size (a.k.a. swap_hdr_sz)
+static size_t
+UnpackPrefix(const char * const buf, const size_t size)
+{
+    Assure(buf);
+    auto input = buf;
+    const auto end = buf + size;
+
+    char magic = 0;
+    SwapMetaExtract(magic, input, end);
+
+    if (magic != SwapMetaMagic)
+        throw TextException("store entry metadata prefix is corrupted", Here());
+
+    RawSwapMetaPrefixLength rawMetaSize = 0; // metadata size, including the required prefix
+    SwapMetaExtract(rawMetaSize, input, end);
+
+    if (Less(rawMetaSize, SwapMetaPrefixSize))
+        throw TextException("store entry metadata length is corrupted", Here());
+
+    return rawMetaSize; // now safe to use within (buf, buf+size)
+}
+
+} // namespace Store
+
+/* Store::SwapMetaIterator */
+
+Store::SwapMetaIterator::SwapMetaIterator(const void * const start, const void * const end):
+    fieldStart_(static_cast<const char*>(start)),
+    bufEnd_(end)
+{
+    sync();
+}
+
+Store::SwapMetaIterator &
+Store::SwapMetaIterator::operator++()
+{
+    Assure(fieldStart_ != bufEnd_);
+    fieldStart_ += sizeof(RawSwapMetaType); // swap meta type
+    fieldStart_ += sizeof(RawSwapMetaLength); // swap meta value length
+    fieldStart_ += meta_.rawLength; // swap meta value
+
+    sync();
+    return *this;
+}
+
+/// (re)set meta_
+void
+Store::SwapMetaIterator::sync()
+{
+    if (fieldStart_ == bufEnd_)
+        return; // nothing to do when we reach the end of iteration
+
+    // We cannot start beyond the end of the header: We start with valid
+    // begin/end buffer pointers, and each field checks for overreach.
+    Assure(fieldStart_ < bufEnd_);
+
+    meta_ = SwapMetaView(fieldStart_, bufEnd_);
+}
+
+/* Store::SwapMetaUnpacker */
+
+Store::SwapMetaUnpacker::SwapMetaUnpacker(const char * const buf, const ssize_t size, size_t &swap_hdr_sz)
+{
+    Assure(buf);
+    Assure(size >= 0);
+
+    const auto headerSize = UnpackPrefix(buf, size);
+    Assure(headerSize >= SwapMetaPrefixSize);
```

Comment:
This check looks redundant becauseUnpackPrefix() has already verified that `headerSize` is less than `SwapMetaPrefixSize`.

Reply:
Ok. Yes, in that unclear phrasing I meant that UnpackPrefix() has checked the `headerSize < SwapMetaPrefixSize` condition.